http/subscribe.go

```diff
@@ -0,0 +1,305 @@
+	h.submtx.Lock()
+	defer h.submtx.Unlock()
+	h.head = c
+	return nil
+}
+
+func (h *httpSubscriber) Sync(ctx context.Context, p peer.ID, c cid.Cid, selector ipld.Node) (chan cid.Cid, context.CancelFunc, error) {
+	respChan := make(chan cid.Cid, 1)
+	cctx, cncl := context.WithCancel(ctx)
+	h.submtx.Lock()
+	defer h.submtx.Unlock()
+	// todo: error if reqs is full
+	h.reqs <- req{
+		Cid:      c,
+		Selector: selector,
+		ctx:      cctx,
+		resp:     respChan,
+	}
+	return respChan, cncl, nil
+}
+
+func (h *httpSubscriber) Close() error {
+	// cancel out subscribers.
+	h.Client.CloseIdleConnections()
+	h.submtx.Lock()
+	defer h.submtx.Unlock()
+	for _, ca := range h.subs {
+		close(ca)
+	}
+	h.subs = make([]chan cid.Cid, 0)
+	return nil
+}
+
+// background event loop for scheduling
+// a. time-scheduled fetches to the provider
+// b. interrupted fetches in response to synchronous 'Sync' calls.
+func (h *httpSubscriber) background() {
+	var nextCid cid.Cid
+	var workResp chan cid.Cid
+	var ctx context.Context
+	var sel ipld.Node
+	var xsel selector.Selector
+	var err error
+	for {
+		select {
+		case r := <-h.reqs:
+			nextCid = r.Cid
+			workResp = r.resp
+			sel = r.Selector
+			ctx = r.ctx
+		case <-time.After(defaultPollTime):
+			nextCid = cid.Undef
+			workResp = nil
+			ctx = context.Background()
+			sel = nil
+		}
+		if nextCid == cid.Undef {
+			nextCid, err = h.fetchHead(ctx)
+		}
+		if sel == nil {
+			sel = h.defaultSelector
+		}
+		if err != nil {
+			log.Warnf("failed to fetch new head: %s", err)
+			err = nil
+			goto next
+		}
+		sel = legs.ExploreRecursiveWithStopNode(selector.RecursionLimitNone(),
+			sel,
+			cidlink.Link{Cid: h.head})
+		if err := h.fetchBlock(ctx, nextCid); err != nil {
+			//log
+			goto next
+		}
+		xsel, err = selector.CompileSelector(sel)
+		if err != nil {
+			//log
+			goto next
+		}
+
+		err = h.walkFetch(ctx, nextCid, xsel)
+		if err != nil {
+			//log
+			goto next
+		}
+
+	next:
+		if workResp != nil {
+			workResp <- nextCid
+			close(workResp)
+			workResp = nil
+		}
+	}
+}
+
+func (h *httpSubscriber) walkFetch(ctx context.Context, root cid.Cid, sel selector.Selector) error {
+	getMissingLs := cidlink.DefaultLinkSystem()
+	getMissingLs.TrustedStorage = true
+	getMissingLs.StorageReadOpener = func(lc ipld.LinkContext, l ipld.Link) (io.Reader, error) {
+		r, err := h.lsys.StorageReadOpener(lc, l)
+		if err == nil {
+			return r, nil
+		}
+		// get.
+		c := l.(cidlink.Link).Cid
+		if err := h.fetchBlock(ctx, c); err != nil {
+			return nil, err
+		}
+		return h.lsys.StorageReadOpener(lc, l)
+	}
+
+	progress := traversal.Progress{
+		Cfg: &traversal.Config{
+			Ctx:                            ctx,
+			LinkSystem:                     getMissingLs,
+			LinkTargetNodePrototypeChooser: basicnode.Chooser,
+		},
+		Path: datamodel.NewPath([]datamodel.PathSegment{}),
+		LastBlock: struct {
+			Path datamodel.Path
+			Link datamodel.Link
+		}{
+			Path: ipld.Path{},
+			Link: nil,
+		},
```

Comment:
Since these are the zero values, the do not need to be explicitly assigned:
```go
	LastBlock: struct {
		Path datamodel.Path
		Link datamodel.Link
	}{},
```